// @ohos/sdkmanager-common/build/src/core/loader/abstract-component-dto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiVersion = exports.MetaDto = exports.AbstractComponentDto = void 0;
const utils_1 = require("../util/utils");
const illegal_apiversion_exception_1 = require("../api/illegal-apiversion-exception");
class AbstractComponentDto {
    constructor() {
        this.displayName = '';
        this.path = '';
        this.releaseType = '';
        this.version = '';
        this.license = '';
    }
    compareTo(other) {
        const comparePath = this.path.localeCompare(other.getPath());
        if (comparePath !== 0) {
            return comparePath;
        }
        const compareApiVersionResult = (0, utils_1.compareApiVersion)(this.fullApiVersion, other.getFullApiVersion());
        if (compareApiVersionResult !== 0) {
            return compareApiVersionResult;
        }
        return (0, utils_1.compareVersion)(this.version, other.getVersion());
    }
    getApiVersion() {
        return -1;
    }
    getDisplayName() {
        return this.displayName;
    }
    getLocation() {
        return this.location;
    }
    getMeta() {
        return this.meta;
    }
    getPath() {
        return this.path;
    }
    getReleaseType() {
        return this.releaseType;
    }
    getVersion() {
        return this.version;
    }
    getFullApiVersion() {
        return this.fullApiVersion;
    }
    hasApiVersion() {
        return (0, utils_1.isValidApiVersion)(this.fullApiVersion);
    }
    getLicense() {
        return this.license;
    }
    getFullName() {
        return `${this.displayName}:${this.version}`;
    }
}
exports.AbstractComponentDto = AbstractComponentDto;
class MetaDto {
    constructor() {
        this.metaVersion = '';
    }
    getMetaVersion() {
        return this.metaVersion;
    }
}
exports.MetaDto = MetaDto;
class ApiVersion {
    constructor(apiVersionStr) {
        this.major = 0;
        this.minor = 0;
        this.patch = 0;
        if (apiVersionStr === undefined || apiVersionStr === null ||
            !(0, utils_1.isValidApiVersionStr)(apiVersionStr)) {
            throw new illegal_apiversion_exception_1.IllegalApiVersionException();
        }
        this.value = apiVersionStr;
        const parts = apiVersionStr.split('.');
        if ((parts === null || parts === void 0 ? void 0 : parts.length) >= 1) {
            this.major = Number(parts[0]);
        }
        if ((parts === null || parts === void 0 ? void 0 : parts.length) >= 2) {
            this.minor = Number(parts[1]);
        }
        if ((parts === null || parts === void 0 ? void 0 : parts.length) >= 3) {
            this.patch = Number(parts[2]);
        }
    }
    compareTo(other) {
        return (0, utils_1.compareApiVersion)(this, other);
    }
    getMajor() {
        return this.major;
    }
    getMinor() {
        return this.minor;
    }
    getPatch() {
        return this.patch;
    }
    getValue() {
        return this.value;
    }
}
exports.ApiVersion = ApiVersion;
