// @ohos/sdkmanager-common/build/src/core/util/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNotEmpty = exports.isEmpty = exports.compareApiVersion = exports.isValidApiVersionStr = exports.isValidApiVersion = exports.isEqualApiVersion = exports.isInMetaVersionList = exports.compareVersion = void 0;
const component_contants_1 = require("../constants/component-contants");
const illegal_apiversion_exception_1 = require("../api/illegal-apiversion-exception");
/**
 * 比较版本号的方法,版本号的格式是以[0-9]+\.的形式组成，例如1.0.0.0
 * 如果字符串不是以上面的规则组成，则直接进行字符串比较
 * @param first
 * @param second
 * @return number 返回结果大于0表示first比second大，
 * 等于0表示相等，小于0表示first比second小
 */
function compareVersion(first, second) {
    const version1 = first.split('.');
    const version2 = second.split('.');
    for (let i = 0; i < version1.length; i++) {
        const number1 = Number(version1[i]);
        const number2 = Number(version2[i]);
        if (Number.isNaN(number1) || Number.isNaN(number2)) {
            return first.localeCompare(second);
        }
        const result = number1 - number2;
        if (result === 0) {
            continue;
        }
        else if (result > 0) {
            return 1;
        }
        else {
            return -1;
        }
    }
    return 0;
}
exports.compareVersion = compareVersion;
/**
* 判断 metaVersion 是否在版本区间内
* @param metaVersion 待检测版本号
* @param metaVersionList 版本号区间（升序排列）
* @returns
* -1: 小于最小值
*  0: 在区间内
*  1: 大于最大值
*/
function isInMetaVersionList(metaVersion, metaVersionList) {
    if (isEmpty(metaVersion)) {
        throw new Error('metaVersion should not be null or empty');
    }
    if (!metaVersionList || metaVersionList.length === 0) {
        throw new Error('metaVersionList should not be null or empty');
    }
    const list = [...metaVersionList];
    list.sort((a, b) => compareVersion(a, b));
    const minVersion = list[0];
    const maxVersion = list[list.length - 1];
    const minCompare = compareVersion(metaVersion, minVersion);
    const maxCompare = compareVersion(metaVersion, maxVersion);
    if (minCompare < 0) {
        return -1;
    }
    if (maxCompare > 0) {
        return 1;
    }
    return 0;
}
exports.isInMetaVersionList = isInMetaVersionList;
/**
 * 比较apiVersion
 *
 * @param apiVersion1
 * @param apiVersion2
 * @return boolean 返回表示两个apiVersion是否相等
 */
function isEqualApiVersion(apiVersion1, apiVersion2) {
    return compareApiVersion(apiVersion1, apiVersion2) === 0;
}
exports.isEqualApiVersion = isEqualApiVersion;
/**
 * 判断apiVersion是否有效
 *
 * @param apiVersion
 * @return boolean 返回是否有效apiVersion
 */
function isValidApiVersion(apiVersion) {
    if (apiVersion === undefined || apiVersion === null) {
        return false;
    }
    const value = apiVersion.getValue();
    if (isEmpty(value)) {
        return false;
    }
    return component_contants_1.API_VERSION_PATTERN.test(value);
}
exports.isValidApiVersion = isValidApiVersion;
/**
 * 判断apiVersion是否有效
 *
 * @param apiVersionStr
 * @return boolean 返回是否有效apiVersion
 */
function isValidApiVersionStr(apiVersionStr) {
    if (apiVersionStr === null || apiVersionStr === undefined || apiVersionStr.length === 0) {
        return false;
    }
    return component_contants_1.API_VERSION_PATTERN.test(apiVersionStr);
}
exports.isValidApiVersionStr = isValidApiVersionStr;
/**
 * 比较apiVersion
 *
 * @param apiVersion1
 * @param apiVersion2
 * @return number 返回结果大于0表示first比second大，
 * 等于0表示相等，小于0表示first比second小
 */
function compareApiVersion(apiVersion1, apiVersion2) {
    if (apiVersion1 === null || apiVersion1 === undefined) {
        throw new illegal_apiversion_exception_1.IllegalApiVersionException('apiVersion1 should not be null or undefined');
    }
    if (apiVersion2 === null || apiVersion2 === undefined) {
        throw new illegal_apiversion_exception_1.IllegalApiVersionException('apiVersion2 should not be null or undefined');
    }
    // 逐级比较主版本、次版本、修订号
    const compareMajor = apiVersion1.getMajor() - apiVersion2.getMajor();
    if (compareMajor !== 0) {
        return compareMajor;
    }
    const compareMinor = apiVersion1.getMinor() - apiVersion2.getMinor();
    if (compareMinor !== 0) {
        return compareMinor;
    }
    return apiVersion1.getPatch() - apiVersion2.getPatch();
}
exports.compareApiVersion = compareApiVersion;
/**
 * 判断值是否为null、undefined、''
 * @param value
 */
function isEmpty(value) {
    return value === null || value === undefined || value.length === 0;
}
exports.isEmpty = isEmpty;
/**
 * 判断值是否为null、undefined、''
 * @param value
 */
function isNotEmpty(value) {
    return !isEmpty(value);
}
exports.isNotEmpty = isNotEmpty;
